### PR TITLE
Allow users to pass in HttpClient for file upload/module method requests

### PR DESCRIPTION
**Pull request by timtay-microsoft**
The implementation already uses the settings for both file upload operations and module HTTP operations, but this now adds the option to pass in an HTTP client.

Similar to #3923, but for v2

### CODE REVIEW

File: iothub/device/src/IotHubBaseClient.cs [hunk 1]
```diff
@@ -73,7 +73,7 @@ public abstract class IotHubBaseClient : IAsyncDisposable
             if (IotHubConnectionCredentials.ModuleId.IsNullOrWhiteSpace())
```
Comment (abhipsaMisra) [inline iothub/device/src/IotHubBaseClient.cs:73]:
```
            PipelineContext.HttpOperationTransportSettings = _clientOptions.HttpOperationTransportSettings;


            if (!IotHubConnectionCredentials.ModuleId.IsNullOrWhiteSpace())
            {
                // Set the remote certificate validator for the module client
                _certValidator = certificateValidator ?? NullCertificateValidator.Instance;
                PipelineContext.HttpOperationTransportSettings.ServerCertificateCustomValidationCallback = _certValidator.GetCustomCertificateValidation();
            }
```

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Transport/Http/HttpClientHelper.cs [hunk 2]
```diff
@@ -65,22 +71,24 @@ internal sealed class HttpClientHelper : IDisposable
...
+                httpClientHandler.Proxy = proxy;
             }
 
-            _httpClientHandler.MaxConnectionsPerServer = DefaultMaxConnectionsPerServer;
+            httpClientHandler.MaxConnectionsPerServer = DefaultMaxConnectionsPerServer;
             ServicePoint servicePoint = ServicePointManager.FindServicePoint(_baseAddress);
             servicePoint.ConnectionLeaseTimeout = s_defaultConnectionLeaseTimeout.Milliseconds;
 
-            _httpClientObj = new HttpClient(_httpClientHandler)
+            _httpMessageHandler = httpClientHandler;
```
Comment (abhipsaMisra) [inline iothub/device/src/Transport/Http/HttpClientHelper.cs:89]: why not do this initialization on L64 directly?

File: iothub/device/src/TransportSettings/IotHubClientHttpSettings.cs [hunk 3]
```diff
@@ -21,6 +21,21 @@ public IotHubClientHttpSettings()
             Protocol = IotHubClientTransportProtocol.WebSocket;
         }
 
+        /// <summary>
+        /// The HTTP client to use for all HTTP operations.
+        /// </summary>
+        /// <remarks>
+        /// If not provided, an HTTP client will be created for you based on the other settings provided.
+        /// <para>
+        /// This HTTP client instance will be disposed when the device/module client using it is disposed.
```
Comment (abhipsaMisra) [inline iothub/device/src/TransportSettings/IotHubClientHttpSettings.cs:30]: The disposal code seems to be missing from this PR. Was it already done as a part of the HttpTransportHandler implementation for the HttpClient that the sdk was creating and now we've substituted the user supplied client for the appropriate case?

File: e2e/Tests/iothub/device/FileUploadE2ETests.cs [hunk 4]
```diff
@@ -100,6 +101,41 @@ public async Task FileUpload_SmallFile_GranularSteps_Proxy()
             await UploadFileGranularAsync(fileStreamSource, filename, fileUploadTransportSettings, isCorrelationIdValid: true, ct: cts.Token).ConfigureAwait(false);
         }
 
+        // File upload requests can be configured to use a user-provided HttpMessageHandler
```
Comment (timtay-microsoft) [inline e2e/Tests/iothub/device/FileUploadE2ETests.cs:104]:
```suggestion
        // File upload requests can be configured to use a user-provided HttpClient
```

File: e2e/Tests/iothub/device/FileUploadE2ETests.cs [hunk 5]
```diff
@@ -100,6 +101,41 @@ public async Task FileUpload_SmallFile_GranularSteps_Proxy()
...
+        [TestMethod]
+        public async Task FileUpload_UsesCustomHttpMessageHandler()
```
Comment (timtay-microsoft) [inline e2e/Tests/iothub/device/FileUploadE2ETests.cs:106]:
```suggestion
        public async Task FileUpload_UsesCustomHttpClient()
```

File: iothub/device/src/Transport/Http/HttpClientHelper.cs (hunk 2 above)
Comment (timtay-microsoft) [reply iothub/device/src/Transport/Http/HttpClientHelper.cs:89]: ```_httpMessageHandler``` is the interface and only the implementation classes have fields for proxy, connection lease timeout, etc.

Comment (timtay-microsoft) [reply iothub/device/src/Transport/Http/HttpClientHelper.cs:89]: Actually, scratch that. That was the case in v1, but here in v2 I can simplify this

File: iothub/device/src/Transport/Http/HttpClientHelper.cs [hunk 6]
```diff
@@ -65,22 +71,24 @@ internal sealed class HttpClientHelper : IDisposable
...
+
+            _httpClientObj = new HttpClient(_httpMessageHandler)
```
Comment (timtay-microsoft) [inline iothub/device/src/Transport/Http/HttpClientHelper.cs:91]:
```suggestion
            _httpClientObj = new HttpClient(_httpMessageHandler, true)
```

File: iothub/device/src/TransportSettings/IotHubClientHttpSettings.cs (hunk 3 above)
Comment (timtay-microsoft) [reply iothub/device/src/TransportSettings/IotHubClientHttpSettings.cs:30]: Correct. ```HttpClientHelper``` was already disposing of the local HttpClient field. This PR just changes how it is instantiated/not instantiated

**Pull request merged**